accepted/2022/FunctionPointers.md

```diff
@@ -0,0 +1,431 @@
+"System.Int32()*[]" | `delegate*<int>*[]`
+"System.Int32()()" | `delegate*<delegate*>int>>`
+"System.Int32()[]" | `delegate*<int>[]`
+"System.Boolean(System.String(System.Int32))" | `delegate*<delegate*<int, string>, bool>`
+
+### `FullName` and `AssemblyQualifiedName` properties
+Returns `null`. This is a simplification that avoids a [type name grammar update](https://docs.microsoft.com/dotnet/framework/reflection-and-codedom/specifying-fully-qualified-type-names) that would need to include grammar for custom modifiers as well as supporting  `Type.GetType()`, `Assembly.GetType()` and any other similar APIs that parse those strings in order to construct a function pointer from it.
+
+Support for a grammar update is possible in the future, with a minimal breaking change here.
+
+Note that this property is already nullable and does return `null` in open generic parameters cases as well.
+
+### `Name` property
+Returns "`*()`".
+
+This property is not nullable.
+
+Consider that generics do not provide parameter type information either:
+```cs
+    string s = typeof(Action<int>).Name; // "Action`1
+```
+
+### `Namespace` property
+Returns `null`.
+
+A standard pointer returns the namespace of the referenced type, but we don't have that with function pointers since they are declared inline.
+
+### `IsPointer` property
+Returns `false`. Per past design discussions, this should be `false` since function pointers and standard pointers are not the same, and this also is non-breaking for existing code that needs to determine if a type is a standard pointer type.
+
+### `IsValueType` and `IsClass` properties
+Use the same behavior as standard pointers which is `IsValueType == false` and `IsClass == true`.
+
+The type hierarchy is covered somewhat [here](https://learn.microsoft.com/previous-versions/visualstudio/visual-studio-2008/2hf02550(v=vs.90)) and in ECMA\CLI: 
+- Type
+  - Value types
+    - Built-in value types
+    - User-defined value types
+    - Enumerations
+  - Reference types
+    - **Pointer types**
+    - Interface types
+    - Self-describing types
+      - Arrays
+      - Class types
+        - User-defined classes
+        - Boxed value types
+        - Delegates
+
+# Proposed APIs
+## System.Type
+```diff
+namespace System
+{
+    public abstract class Type
+    {
+        // These throw InvalidOperationException if IsFunctionPointer is false.
++       public virtual Type GetFunctionPointerReturnType();
++       public virtual Type[] GetFunctionPointerParameterTypes();
++       public virtual bool IsFunctionPointer { get; }
+
+        // Required since this can't be determined by inspecting the results from GetFunctionPointerCallingConventions()
++       public virtual bool IsUnmanagedFunctionPointer { get; }
+
+        // Exposing here without requiring a signature type is debatable; it is here since this
+        // information may be useful for runtime scenarios, and not just metadata inspection.
```

Comment:
It is hard to tell whether it is more likely for runtime scenarios to need calling convention modopts as opposed other modopts.

Reply:
The design now says GetCallingConventions does not work unless it is for a signature type.

However, the existing implementation for CoreClr uses the "CallKind" in various locations, so that does introduce some equality concerns -- other compilers than C# may not use "CallKind" and instead may use mod opts.